src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -227,6 +237,814 @@ class OMPSerialDirective(OMPRegionDirective, metaclass=abc.ABCMeta):
+                    sorted_dependency_pairs = (list(t) for t in
+                                               zip(*sorted(zip(
+                                                   highest_position_nodes,
+                                                   lowest_position_nodes,
+                                                   dependent_nodes))))
+        else:
+            # We have no invalid dependencies so we can return early
+            return
+
+        taskwait_location_nodes = []
+        taskwait_location_abs_pos = []
+        for taskwait in self.walk(OMPTaskwaitDirective):
+            taskwait_location_nodes.append(taskwait)
+            taskwait_location_abs_pos.append(taskwait.abs_position)
+        # Add the first node to have a taskwait placed in front of it into the
+        # list
+        taskwait_location_nodes.append(sorted_dependency_pairs[0][1])
+        taskwait_location_abs_pos.append(sorted_highest_positions[0])
+        for index, pairs in enumerate(sorted_dependency_pairs[1:]):
+            # Add 1 to index here because we're looking from [1:]
+            lo_abs_pos = sorted_lowest_positions[index+1]
+            hi_abs_pos = sorted_highest_positions[index+1]
+            satisfied = False
+            for ind, taskwait_loc in enumerate(taskwait_location_nodes):
+                if (taskwait_location_abs_pos[ind] <= hi_abs_pos and
+                        taskwait_location_abs_pos[ind] >= lo_abs_pos):
+                    # We have a taskwait meant to be placed here that is
+                    # potentially already satisfied. To check we need to
+                    # ensure that the ancestor schedules of the nodes
+                    # are identical
+                    if (pairs[0].ancestor(Schedule) is
+                            taskwait_loc.ancestor(Schedule)):
+                        satisfied = True
+                        break
+            # If we didn't find a taskwait we plan to add that satisfies this
+            # dependency, add it to the list
+            if not satisfied:
+                taskwait_location_nodes.append(pairs[1])
+                taskwait_location_abs_pos.append(hi_abs_pos)
+        # Now loop through the list in reverse and add taskwaits
+        taskwait_location_nodes.reverse()
+        for taskwait_loc in taskwait_location_nodes:
+            node_parent = taskwait_loc.parent
+            loc = taskwait_loc.position
+            node_parent.addchild(OMPTaskwaitDirective(), loc)
+            # As the node_parent tree has already been lowered and we're adding
+            # a new node, we need to lower it again.
+            node_parent.lower_to_language_level()
+
+    def lower_to_language_level(self):
+        '''
+        Checks that any task dependencies inside this node are valid.
+        '''
+        # Perform parent ops
+        super().lower_to_language_level()
+
+        # Check task dependencies are satisfiable.
```

Comment:
Now that I get to this point, I realise that calling this `check_xxx` implies that that's all it does. However, it now
adds OMPTaskwaitDirectives as required so I think it needs a name change (and the comment should reflect
this too).

Reply:
I've changed this to `_validate_task_dependencies` and adjusted the documentation.